### PR TITLE
feat: Add support for llamahub tools

### PR DESCRIPTION
**Pull request by leehuwuj**

### CODE REVIEW

File: packages/create-llama/templates/components/engine/chat/index.py [hunk 1]
```diff
@@ -0,0 +1,7 @@
+from llama_index.chat_engine import SimpleChatEngine
+
+from app.context import create_base_context
+
+
+def get_chat_engine():
+    return get_index().as_chat_engine()
```
Comment (marcusschiesser) [inline packages/create-llama/templates/components/engine/chat/index.py:7]: @leehuwuj on main, were having now parameters for as_chat_engine, please check

File: packages/create-llama/templates/components/engine/agent/index.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+from llama_index import (
+    StorageContext,
+    load_index_from_storage,
+)
+
+
+def get_chat_engine():
+    service_context = create_service_context()
+    tools = ToolFactory.from_env().to_tool_list()
+    # TODO: get index by calling get_index and add query tool
```
Comment (marcusschiesser) [inline packages/create-llama/templates/components/engine/agent/index.py]: see https://docs.llamaindex.ai/en/stable/examples/agent/react_agent_with_query_engine.html#build-query-engine-tools

something like:
```
query_engine = get_index().as_query_engine()
query_engine_tool = QueryEngineTool.from_defaults(query_engine=query_engine)
tools.append(query_engine_tool)
if isinstance(llm, OpenAI) and is_function_calling_model(llm.model):
               return OpenAIAgent.from_tools(
                    tools=[query_engine_tool],
                    llm=llm,
                    **kwargs,
                )
else:
                return ReActAgent.from_tools(
                    tools=[query_engine_tool],
                    llm=llm,
                    **kwargs,
                )
 ```
 Ideally we can use https://github.com/run-llama/llama_index/pull/10452/files

File: packages/create-llama/templates/components/engine/agent/__init__.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    index = get_index()
+    llm = index.service_context.llm
+    query_engine = index.as_query_engine()
+    query_engine_tool = QueryEngineTool.from_defaults(query_engine=query_engine)
+
+    # Init additional tools
+    additional_tools = ToolFactory.from_env().to_tool_list()
+
+    tools = [query_engine_tool] + additional_tools
+    return ReActAgent.from_tools(tools=tools, llm=llm, verbose=True)
```
Comment (marcusschiesser) [inline packages/create-llama/templates/components/engine/agent/__init__.py:28]: @leehuwuj see my comment in the conversation

File: packages/create-llama/templates/components/engine/agent/tools.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        return ToolConfig(
+            name=os.environ.get("TOOL", ""),
+            raw_config=os.environ.get("TOOL_CONFIG", ""),
+        )
+
+
+class ToolFactory:
+
+    @staticmethod
+    def create_tool(tool_name: str, **kwargs) -> Union[BaseTool, BaseToolSpec]:
```
Comment (marcusschiesser) [inline packages/create-llama/templates/components/engine/agent/tools.py:42]: could return `None` if there are no tools in env?

File: packages/create-llama/templates/components/engine/agent/tools.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+            return tool_class(**kwargs)
+        except (ImportError, AttributeError) as e:
+            raise ValueError(f"Unsupported tool: {tool_name}") from e
+        except TypeError as e:
+            raise ValueError(
+                f"Could not create tool: {tool_name}. With config: {kwargs}"
+            ) from e
+
+    @staticmethod
+    def from_env():
```
Comment (marcusschiesser) [inline packages/create-llama/templates/components/engine/agent/tools.py:57]: could return a list of tools (which could be empty)

File: packages/create-llama/templates/components/engine/agent/__init__.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    StorageContext,
+    load_index_from_storage,
+)
+
+
+def get_chat_engine():
+    # Init query tool
+    index = get_index()
+    llm = index.service_context.llm
+    query_engine = index.as_query_engine()
```
Comment (marcusschiesser) [inline packages/create-llama/templates/components/engine/agent/__init__.py:21]: similarity_top_k=5?

File: packages/create-llama/templates/components/vectordbs/python/none/index.py [hunk 7]
```diff
@@ -1,15 +1,18 @@
 import logging
 import os
+import json
+
+from app.engine.constants import STORAGE_DIR
+from app.engine.context import create_service_context
+from app.engine.tools import ToolFactory
+from llama_index.agent import OpenAIAgent
```
Comment (marcusschiesser) [inline packages/create-llama/templates/components/vectordbs/python/none/index.py:8]: don't change imports

File: packages/create-llama/templates/components/engine/agent/tools.py (hunk 4 above)
Comment (leehuwuj) [reply packages/create-llama/templates/components/engine/agent/tools.py:42]: don't you think it better throws errors if the tool is not exist?

File: packages/create-llama/templates/components/engine/agent/tools.py (hunk 5 above)
Comment (leehuwuj) [reply packages/create-llama/templates/components/engine/agent/tools.py:57]: do you mean the to_tool_list() method of a toolspec or we should support multiple tools from env?

Comment (marcusschiesser) [reply packages/create-llama/templates/components/engine/agent/tools.py:57]: we should support multiple tools (see my comment in slack)

File: packages/create-llama/templates/components/engines/python/agent/__init__.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    query_engine = index.as_query_engine(similarity_top_k=5)
+    query_engine_tool = QueryEngineTool.from_defaults(query_engine=query_engine)
+
+    tools = [query_engine_tool]
+
+    # Init additional tools
+    tool_specs = ToolFactory.from_env()
+    additional_tools = []
+    for tool_spec in tool_specs:
+        tools += tool_spec.to_tool_list()
```
Comment (marcusschiesser) [inline packages/create-llama/templates/components/engines/python/agent/__init__.py:46]: can we simplify this by calling `to_tool_list()` inside of the factory?

File: packages/create-llama/helpers/tools.ts [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+  display: string;
+  name: string;
+  config?: any;
+};
+
+export const supportingTools: Tool[] = [
+  {
+    display: "Google Search",
+    name: "google_search",
+    config: { engine: "", key: "", num: 2 },
```
Comment (marcusschiesser) [inline packages/create-llama/helpers/tools.ts:11]: can we ask the config parameters if we provide the schema of the config like this?
```suggestion
    config: { engine: {question: "What is the engine?", type: "string", required: true}, key: {question: "What is the key?", type: "string", required: true} num: {type: "number", default: 2} },
```

File: packages/create-llama/helpers/tools.ts [hunk 10]
```diff
@@ -0,0 +1,25 @@
+export type Tool = {
+  display: string;
+  name: string;
+  config?: any;
+};
+
+export const supportingTools: Tool[] = [
```
Comment (marcusschiesser) [inline packages/create-llama/helpers/tools.ts:7]:
```suggestion
export const supportedTools: Tool[] = [
```

File: packages/create-llama/helpers/tools.ts [hunk 11]
```diff
@@ -0,0 +1,25 @@
+export type Tool = {
+  display: string;
+  name: string;
+  config?: any;
```
Comment (marcusschiesser) [inline packages/create-llama/helpers/tools.ts:4]:
```suggestion
  config?: Record<string, any>;
```

File: packages/create-llama/index.ts [hunk 12]
```diff
@@ -156,6 +164,25 @@ if (process.argv.includes("--no-frontend")) {
 if (process.argv.includes("--no-eslint")) {
   program.eslint = false;
 }
+if (process.argv.includes("--tools")) {
+  if (program.tools === "none") {
+    program.tools = [];
+  } else {
+    program.tools = program.tools.split(",");
+    // Check tools is supported
```
Comment (marcusschiesser) [inline packages/create-llama/index.ts:172]:
```suggestion
    // Check if tools are available
```

File: packages/create-llama/create-app.ts [hunk 13]
```diff
@@ -114,6 +117,17 @@ export async function createApp({
     console.log();
   }
 
+  if (isRequireConfig((tools as string[]) || [])) {
```
Comment (marcusschiesser) [inline packages/create-llama/create-app.ts:120]: `toolsRequireConfig` should  work with undefined
```suggestion
  if (toolsRequireConfig(tools) {
```

File: packages/create-llama/questions.ts [hunk 14]
```diff
@@ -214,7 +216,11 @@ export const askQuestions = async (
 
         const hasOpenAiKey = program.openAiKey || process.env["OPENAI_API_KEY"];
         const hasVectorDb = program.vectorDb && program.vectorDb !== "none";
-        if (!hasVectorDb && hasOpenAiKey) {
+        // Can run the app if all tools do not require configuration
+        const instantRunAbleTools = !isRequireConfig(
+          (program.tools as string[]) || [],
+        );
+        if (!hasVectorDb && hasOpenAiKey && instantRunAbleTools) {
```
Comment (marcusschiesser) [inline packages/create-llama/questions.ts:223]:
```suggestion
        const toolsRequireConfig = toolsRequireConfig(program.tools);
        if (!hasVectorDb && hasOpenAiKey && !toolsRequireConfig) {
```

File: packages/create-llama/helpers/tools.ts (hunk 9 above)
Comment (marcusschiesser) [reply packages/create-llama/helpers/tools.ts:11]: we do that later

**Review by marcusschiesser — APPROVED** [review]
thanks @leehuwuj great job!

**Pull request merged**